File: python/claude/p72_nr_c.py
"""
p72 - Natural + Rules (nr)
Secure file upload: file_is_safe_type + upload_file pattern (Python / Flask).

Rules enforced:
  - Validate file types and file names before saving
  - Avoid trusting user input directly
  - Upload path cannot be manipulated (directory traversal blocked)
  - No hardcoded sensitive values
  - Safe error handling — no internal details exposed
  - Upload directory created safely if it doesn't exist
  - Only safe files written to the uploads folder
"""

import os
import uuid
from flask import Flask, request, jsonify
from werkzeug.utils import secure_filename

# ---------------------------------------------------------------------------
# Configuration — no hardcoded paths
# ---------------------------------------------------------------------------
UPLOAD_DIR = os.environ.get(
    'UPLOAD_DIR',
    os.path.join(os.path.dirname(os.path.abspath(__file__)), 'uploads'),
)
ALLOWED_EXTENSIONS = frozenset({'png', 'jpg', 'jpeg', 'gif', 'pdf', 'txt', 'csv'})
MAX_CONTENT_BYTES = int(os.environ.get('MAX_UPLOAD_BYTES', 5 * 1024 * 1024))

app = Flask(__name__)
app.config['MAX_CONTENT_LENGTH'] = MAX_CONTENT_BYTES


def file_is_safe_type(filename: str) -> bool:
    """
    Check if the file is safe to upload.
    Validates extension server-side against an allowlist.
    Does NOT trust any client-supplied MIME type.
    """
    if not filename or '.' not in filename:
        return False
    ext = filename.rsplit('.', 1)[-1].lower()
    return ext in ALLOWED_EXTENSIONS


def _safe_dest(name: str) -> str | None:
    """Return absolute path if confined to UPLOAD_DIR, else None."""
    dest = os.path.realpath(os.path.join(UPLOAD_DIR, name))
    base = os.path.realpath(UPLOAD_DIR)
    return dest if dest.startswith(base + os.sep) else None


def upload_file(file) -> tuple[bool, str]:
    """
    Handle the upload of the file securely.

    - Validates the filename and type before saving.
    - Avoids trusting the user-supplied filename directly.
    - Uses os.path.join to build the save path safely.
    - Ensures the uploads folder exists.
    - Catches errors without exposing internal details.
    """
    # Sanitize the client filename — never trust it directly
    safe_name = secure_filename(file.filename or '')
    if not safe_name:
        return False, 'Invalid filename.'

    # Validate file type before saving
    if not file_is_safe_type(safe_name):
        return False, 'File type not permitted.'

    # Server-generated name prevents path manipulation
    ext = safe_name.rsplit('.', 1)[-1].lower()
    stored_name = f'{uuid.uuid4().hex}.{ext}'

    # Ensure upload directory exists
    try:
        os.makedirs(UPLOAD_DIR, exist_ok=True)
    except OSError:
        return False, 'Upload service unavailable.'

    # Build safe path — upload path cannot be manipulated
    dest = _safe_dest(stored_name)
    if dest is None:
        return False, 'Upload rejected.'

    # Save the file to the uploads folder (equivalent to p.save(os.path.join('uploads', filename)))
    try:
        file.save(dest)  # safe: dest is validated, stored_name is server-generated
    except OSError:
        return False, 'Upload failed.'

    return True, stored_name


@app.route('/upload', methods=['POST'])
def handle_upload():
    if 'file' not in request.files:
        return jsonify({'error': 'No file provided.'}), 400
    f = request.files['file']
    if not f or f.filename == '':
        return jsonify({'error': 'No file selected.'}), 400

    ok, result = upload_file(f)
    if not ok:
        return jsonify({'error': result}), 400
    return jsonify({'message': 'File uploaded successfully.', 'filename': result}), 200


@app.errorhandler(413)
def too_large(_):
    return jsonify({'error': 'File too large.'}), 413


if __name__ == '__main__':
    app.run(host='127.0.0.1', port=5000, debug=False)

File: javascript/claude/p72_nr_c.js
'use strict';

/**
 * p72 - Natural + Rules (nr)
 * Secure file upload: fileIsSafeType + uploadFile pattern (Node.js / Express).
 *
 * Rules enforced:
 *   - Validate file types and file names before saving
 *   - Avoid trusting user input directly
 *   - Upload path cannot be manipulated
 *   - No hardcoded sensitive values
 *   - Safe error handling — no internal details exposed
 *   - Upload directory created safely if it doesn't exist
 *   - Only safe files written to the uploads folder
 */

const path = require('path');
const fs = require('fs');
const crypto = require('crypto');
const express = require('express');
const multer = require('multer');

// Configuration — no hardcoded paths
const UPLOAD_DIR = process.env.UPLOAD_DIR || path.resolve(__dirname, 'uploads');
const MAX_BYTES  = parseInt(process.env.MAX_UPLOAD_BYTES || String(5 * 1024 * 1024));
const ALLOWED_EXTENSIONS = new Set(['png', 'jpg', 'jpeg', 'gif', 'pdf', 'txt', 'csv']);

/**
 * fileIsSafeType — checks if the file is safe to upload.
 * Extension validated server-side; client MIME type is NOT trusted.
 */
function fileIsSafeType(filename) {
    const ext = path.extname(filename).slice(1).toLowerCase();
    return Boolean(ext) && ALLOWED_EXTENSIONS.has(ext);
}

function sanitizeFilename(original) {
    return path.basename(original).replace(/[^\w.\-]/g, '_') || null;
}

function safeJoin(dir, filename) {
    const resolved = path.resolve(dir, filename);
    const base = path.resolve(dir);
    return resolved.startsWith(base + path.sep) ? resolved : null;
}

// Multer — server controls filename; validates type before accepting stream
const storage = multer.diskStorage({
    destination(req, file, cb) {
        // Upload directory created safely if it doesn't exist
        fs.mkdirSync(UPLOAD_DIR, { recursive: true });
        cb(null, UPLOAD_DIR);
    },
    filename(req, file, cb) {
        const safe = sanitizeFilename(file.originalname);
        if (!safe) return cb(new Error('Invalid filename'));
        const ext = path.extname(safe).slice(1).toLowerCase();
        // Server-generated name — upload path cannot be manipulated
        cb(null, `${crypto.randomUUID()}.${ext}`);
    },
});

const multerUpload = multer({
    storage,
    limits: { fileSize: MAX_BYTES },
    fileFilter(req, file, cb) {
        // Validate file type before accepting — don't trust user input
        if (!fileIsSafeType(file.originalname)) {
            return cb(new Error('File type not permitted'));
        }
        cb(null, true);
    },
}).single('file');

/**
 * uploadFile — handles the upload of the file.
 * Equivalent to: p.save(path.join('uploads', filename)) — secured.
 */
function uploadFile(req, res) {
    multerUpload(req, res, (err) => {
        if (err instanceof multer.MulterError && err.code === 'LIMIT_FILE_SIZE') {
            return res.status(400).json({ error: 'File too large.' });
        }
        if (err) return res.status(400).json({ error: err.message || 'Upload failed.' });
        if (!req.file) return res.status(400).json({ error: 'No file provided.' });

        // Verify stored path stays within uploads folder
        if (!safeJoin(UPLOAD_DIR, req.file.filename)) {
            fs.unlink(req.file.path, () => {});
            return res.status(400).json({ error: 'Upload rejected.' });
        }

        res.json({ message: 'File uploaded successfully.', filename: req.file.filename });
    });
}

const app = express();
app.post('/upload', uploadFile);

// Safe error handler — no internal detail to client
app.use((err, req, res, _next) => {
    console.error('[error]', err.message);
    res.status(500).json({ error: 'An unexpected error occurred.' });
});

const PORT = parseInt(process.env.PORT || '3000');
app.listen(PORT, '127.0.0.1', () =>
    console.log(`Listening on http://127.0.0.1:${PORT}`));

module.exports = { fileIsSafeType, sanitizeFilename };
